File: selenium/TC-func-REQ-SAF-001-4.java
import org.openqa.selenium.*;
import org.openqa.selenium.chrome.ChromeDriver;
import org.openqa.selenium.support.ui.*;
import org.testng.Assert;
import org.testng.annotations.*;
import java.time.Duration;
import java.util.List;

public class TC-func-REQ-SAF-001-4_Test {
    private WebDriver driver;
    private WebDriverWait wait;

    @BeforeMethod
    public void setUp() {
        System.setProperty("webdriver.chrome.driver", "/path/to/chromedriver");
        driver = new ChromeDriver();
        wait = new WebDriverWait(driver, Duration.ofSeconds(10));
        driver.get("http://example.com");
    }

    @Test
    public void test_TC-func-REQ-SAF-001-4() {
        // Step 1: System reads fuel quantity -5 liters
        WebElement fuelInput = wait.until(ExpectedConditions.visibilityOfElementLocated(By.cssSelector("[data-testid='TODO']")));
        fuelInput.clear();
        fuelInput.sendKeys("-5");
        Assert.assertEquals(fuelInput.getAttribute("value"), "-5", "Fuel quantity recorded as -5 liters");

        // Step 2: System detects out-of-range fuel quantity
        WebElement validationMsg = wait.until(ExpectedConditions.visibilityOfElementLocated(By.cssSelector("[data-testid='TODO']")));
        Assert.assertTrue(validationMsg.getText().contains("invalid"), "System flags value as invalid");

        // Step 3: System sets fuel quantity to 0 liters for calculation
        WebElement calculateButton = wait.until(ExpectedConditions.elementToBeClickable(By.cssSelector("[data-testid='TODO']")));
        calculateButton.click();

        WebElement displayedFuel = wait.until(ExpectedConditions.visibilityOfElementLocated(By.cssSelector("[data-testid='TODO']")));
        Assert.assertEquals(displayedFuel.getText(), "0", "Fuel quantity set to 0 liters");

        // Step 4: System calculates estimated driving range
        WebElement rangeElement = wait.until(ExpectedConditions.visibilityOfElementLocated(By.cssSelector("[data-testid='TODO']")));
        Assert.assertEquals(rangeElement.getText(), "0 km", "Estimated range calculated as 0 km");

        // Step 5: System displays estimated driving range
        // Already verified above

        // Step 6: System checks low-fuel warning threshold
        List<WebElement> warningElements = driver.findElements(By.cssSelector("[data-testid='TODO']"));
        Assert.assertTrue(warningElements.isEmpty(), "No warning triggered because range is 0 km");
    }

    @AfterMethod
    public void tearDown() {
        if (driver != null) {
            driver.quit();
        }
    }
}

File: playwright/TC-func-REQ-SAF-001-4.ts
import { test, expect } from '@playwright/test';

test("TC-func-REQ-SAF-001-4 - Validate system handling of sensor out-of-range fuel quantity", async ({ page }) => {
  // Precondition: navigate to the fuel sensor page
  await page.goto('/fuel-sensor');

  // Step 1: System reads fuel quantity -5 liters
  await page.locator('[data-testid="fuel-sensor-input"]').fill('-5');
  await page.locator('[data-testid="fuel-sensor-submit"]').click();

  // Expected: Fuel quantity recorded as -5 liters
  const recordedFuel = await page.locator('[data-testid="fuel-quantity-display"]').textContent();
  expect(recordedFuel).toBe('-5');

  // Step 2: System detects out-of-range fuel quantity
  const errorMessage = await page.locator('[data-testid="fuel-error"]').textContent();
  expect(errorMessage).toContain('invalid');

  // Step 3: System sets fuel quantity to 0 liters for calculation
  const adjustedFuel = await page.locator('[data-testid="fuel-quantity-display"]').textContent();
  expect(adjustedFuel).toBe('0');

  // Step 4: System calculates estimated driving range
  const estimatedRange = await page.locator('[data-testid="estimated-range"]').textContent();
  expect(estimatedRange).toBe('0 km');

  // Step 5: System displays estimated driving range
  const displayedRange = await page.locator('[data-testid="estimated-range-display"]').textContent();
  expect(displayedRange).toBe('0 km');

  // Step 6: System checks low-fuel warning threshold
  const warningVisible = await page.locator('[data-testid="low-fuel-warning"]').isVisible();
  expect(warningVisible).toBe(false);
});